lib/floe/workflow/choice_rule/data.rb

```diff
@@ -40,52 +56,104 @@ def true?(context, input)
             lhs >= rhs
           when "StringMatches"
             lhs.match?(Regexp.escape(rhs).gsub('\*', '.*?'))
-          else
-            raise Floe::InvalidWorkflowError, "Invalid choice [#{compare_key}]"
           end
         end
 
         private
 
-        def validate!(value)
-          raise "No such variable [#{variable}]" if value.nil? && !%w[IsNull IsPresent].include?(compare_key)
+        def presence_check(context, input)
+          rhs = compare_value(context, input)
+          # don't need the value, just need to see if the path finds the value
+          variable_value(context, input)
+
+          # path found the variable_value, (so if they said true, return true)
+          rhs
+        rescue Floe::PathError
+          # variable_value (path) threw an error
+          # it was not found (so if they said false, return true)
+          !rhs
         end
 
-        def is_null?(value) # rubocop:disable Naming/PredicateName
-          value.nil?
+        # rubocop:disable Naming/PredicateName
+        # rubocop:disable Style/OptionalBooleanParameter
+        def is_null?(value, ret_value = true)
+          value.nil? == ret_value
         end
 
-        def is_present?(value) # rubocop:disable Naming/PredicateName
-          !value.nil?
+        # if it got here (and value was fetched), then it is present.
+        def is_present?(_value, ret_value = true)
+          ret_value
         end
 
-        def is_numeric?(value) # rubocop:disable Naming/PredicateName
-          value.kind_of?(Integer) || value.kind_of?(Float)
+        def is_numeric?(value, ret_value = true)
+          value.kind_of?(Numeric) == ret_value
         end
 
-        def is_string?(value) # rubocop:disable Naming/PredicateName
-          value.kind_of?(String)
+        def is_string?(value, ret_value = true)
+          value.kind_of?(String) == ret_value
         end
 
-        def is_boolean?(value) # rubocop:disable Naming/PredicateName
-          [true, false].include?(value)
+        def is_boolean?(value, ret_value = true)
+          [true, false].include?(value) == ret_value
         end
 
-        def is_timestamp?(value) # rubocop:disable Naming/PredicateName
+        def is_timestamp?(value, ret_value = true)
           require "date"
 
           DateTime.rfc3339(value)
-          true
+          ret_value
         rescue TypeError, Date::Error
-          false
+          !ret_value
         end
+        # rubocop:enable Naming/PredicateName
+        # rubocop:enable Style/OptionalBooleanParameter
 
-        def compare_key
-          @compare_key ||= payload.keys.detect { |key| key.match?(/^(IsNull|IsPresent|IsNumeric|IsString|IsBoolean|IsTimestamp|String|Numeric|Boolean|Timestamp)/) }
+        def variable_value(context, input)
+          fetch_path("Variable", variable, context, input)
         end
```

Comment:
(moved over from ChoiceRule since this is data specific)
Before:
- convert `@variable` to a `Path` at runtime and fetch value.
After:
- `@variable is a `Path`, so bad Path detected at initialization time
- Path not present in input is a runtime error
- Data type of input is checked.